addons/point_of_sale/static/src/app/store/pos_store.js

```diff
@@ -236,13 +236,46 @@ export class PosStore extends Reactive {
         this.config.iface_printers = !!this.unwatched.printers.length;
 
         // Monitor product pricelist
+        this.models["product.product"].addEventListener(
+            "create",
+            this.computeProductPricelistCache.bind(this)
+        );
+        this.models["product.pricelist.item"].addEventListener(
+            "create",
+            this.computeProductPricelistCache.bind(this)
+        );
+
         this.computeProductPricelistCache();
     }
 
-    computeProductPricelistCache() {
+    computeProductPricelistCache(data) {
```

Comment:
make this accept the loaded products and pricelist items and perform the necessary linking.
The linking should only be done on those given values.

Then, explicitly call this method where it is needed. Remove the 'events' mechanism that you introduced. We'll think of proper API for that later.